### ISSUE TITLE
How to adding `--invalid-before` option when buiding tnx for a vesting script?

### ISSUE THREAD

**Issue by ahaxu** [Berry-Pool/lucid#64]
From this vesting [script](https://github.com/input-output-hk/plutus-pioneer-program/blob/main/code/week03/src/Week03/Parameterized.hs).
[This line](https://github.com/input-output-hk/plutus-pioneer-program/blob/main/code/week03/src/Week03/Parameterized.hs#L58) is the on chain part to validate whether deadline is reached.

```
    deadlineReached = contains (from $ deadline p) $ txInfoValidRange info
```
This is the offchain part using lucid
```
export const redeemUtxo = async (plutusScript, number) => {
    const vestingScript = {
       type: 'PlutusV1',
       script: plutusScript,
    };

    const vestingAddress = lucid.utils.validatorToAddress(
              vestingScript
        );

      const utxo = (await lucid.utxosAt(vestingAddress)).slice(-1)[0];

      const tx = await lucid
        .newTx()
        .collectFrom([utxo], Redeemer(number))
        .attachSpendingValidator(vestingScript)
        .addSigner(benificialAddress)
        .complete()

      const signedTx = await tx.sign().complete();

      const txHash = await signedTx.submit();

      return txHash;
};
```
However, **its always failed when building tnx to redeem the vesting script**, even i set the deadline to a specified time which less than the redeem time.
```
localhost/:1 Uncaught (in promise) {
  "ScriptFailures": {
    "spend:0": [
      {
        "validatorFailed": {
          "error": "An error has occurred:  User error:\nThe machine terminated because of an error, either from a built-in function or from an explicit use of 'error'.",
          "traces": [
            "deadline not reached",
            "PT5"
          ]
        }
      }
    ]
  }
}
```
I saw that there is work around for this case which is adding `--invalid-before 48866954` in `cardano-cli transaction build`
https://github.com/input-output-hk/plutus-pioneer-program/blob/main/code/week03/testnet/grab.sh#L11
can we do the same with lucid when building tnx ? 

Thank you!

**Comment by alessandrokonrad** [comment]
Yes! Simply add `validTo(unixTime)` or `validFrom(unixTime)` to your tx if you want to make use of time.
For your tx it looks like you have to add validFrom because that should be equivalent to `--invalid-before`. 
Note Lucid wants you to add the unixTime in milliseconds and not the slot. For instance you could make use of `Date.now()` in Js:
```
      const tx = await lucid
        .newTx()
        .collectFrom([utxo], Redeemer(number))
        .validFrom(...)
        .attachSpendingValidator(vestingScript)
        .addSigner(benificialAddress)
        .complete()
```

> I saw that there is work around for this case which is adding --invalid-before 48866954 in cardano-cli transaction build

It's not a workaround. Adding an interval is necessary if you want your script to read out a certain valid time interval, for Plutus and Native scripts :)

**Comment by ahaxu** [comment]
Thank you so much @alessandrokonrad.

> It's not a workaround. Adding an interval is necessary if you want your script to read out a certain valid time interval, for Plutus and Native scripts :)

can you explain more on this, i really still feel dump why we need to add `--invalid-before`.
from what i understand, this part on on-chain
```
    deadlineReached = contains (from $ deadline p) $ txInfoValidRange info
```
assumed to check if the deadline reached or not.

Can you elaborate more on this ;)

**Comment by ahaxu** [comment]
So, after thinking for a while,
i guess `--invalid-before` or `validFrom` option is checked when script is evaluated (when building **tnx**), before  **tnx** actually hit (submit) **on-chain** logic.

**Comment by ahaxu** [comment]
I watched [this](https://youtu.be/ABtffZPoUqU?t=1764) again (started at **29:24**) slowly (0.5 x**speed**), and everything clear now.
Thank you again!
